src/serializer/ResidualHeapSerializer.js

```diff
@@ -479,29 +481,18 @@ export class ResidualHeapSerializer {
       }
     }
 
-    if (generators.length === 1 && functionValues.length === 0) {
-      // This value is only referenced from a single generator, and it's not the realm generator.
-      invariant(generators[0] !== this.generator);
-      // We can emit the initialization of this value into the body associated with that generator.
-      let body = this.activeGeneratorBodies.get(generators[0]);
-      if (body === undefined) {
-        // TODO: The generator is no longer active! We missed the right time to emit this value.
-        this.logger.logError(val, "Value is referenced in an unsupported scope.");
-        return { body: this.mainBody };
-      } else {
-        invariant(body === this.emitter.getBody());
-        return { body };
-      }
+    // This value is referenced from more than one generator or function.
+    // We can emit the initialization of this value into the body associated with their common ancestor.
+    let commonAncestor = scopeList.reduce((x, y) => x && commonAncestorOf(x, y), generators[0]);
+    invariant(commonAncestor instanceof Generator);
+    let body = commonAncestor === this.generator ? this.mainBody : this.activeGeneratorBodies.get(commonAncestor);
+    if (body === undefined) {
+      // TODO: The generator is no longer active! We missed the right time to emit this value.
```

Comment:
I'm happy to replace this with an invariant if you believe it should never happen.